bindings/rust/src/lib.rs

```diff
@@ -1107,25 +1020,22 @@ macro_rules! sig_variant_impl {
 
                 // TODO - check msg uniqueness?
 
-                let pool = mt::da_pool();
-                let (tx, rx) = channel();
-                let counter = Arc::new(AtomicUsize::new(0));
-                let valid = Arc::new(AtomicBool::new(true));
-
-                let n_workers = core::cmp::min(pool.max_count(), n_elems);
-                for _ in 0..n_workers {
-                    let tx = tx.clone();
-                    let counter = counter.clone();
-                    let valid = valid.clone();
+                let counter = AtomicUsize::new(0);
+                let valid = AtomicBool::new(true);
+                let acc = Mutex::new(None::<Pairing>);
+                let mut gtsig = blst_fp12::default();
 
-                    pool.joined_execute(move || {
+                rayon::scope(|scope| {
+                    scope.spawn_broadcast(|_scope, _ctx| {
+                        let mut processed = 0;
                         let mut pairing = Pairing::new($hash_or_encode, dst);
 
-                        while valid.load(Ordering::Relaxed) {
+                        while valid.load(Ordering::Acquire) {
```

Comment:
`relaxed` ordering is OK there and will be faster on weak memory models like ARM

Reply:
For counter it is OK, for `valid` I'm not so sure. Atomics are hard to get right and with `Relaxed` all kinds of unexpected things can happen with instruction reordering. If we expect to load something after we have written something prior to this, we should use proper ordering here. It doesn't matter on x86-64 to the best of my knowledge for integers/booleans, but I see no reason to not use it to have hard guarantees of correctness on all platforms.

Can you elaborate why it is 100% guaranteed that relaxed ordering here is correct?

Reply:
In that specific `rayon::scope` what matters is that eventually `valid` will be written false iff there is a wrong aggregation.
Even if there is a spurious true because we used `relaxed` order and cache were not synchronized, we just do an extra loop iteration (or 2 or n). But eventually it will be written.

Now what matters is that before line 1079: https://github.com/supranational/blst/pull/203/files/d981fc9baa85be6886f344bfd652f084bffb879b#diff-7f53e1ecce5b519e9002e448e51f119bda7092059ce9b57e9ff3fa1e33767995R1079, which will early return if there is an error, a `valid == false` has been properly propagated.
For full correctness, we can guarantee that with a fence with sequential consistency. But I would assume it it implied by the end of rayon::scope.

Reply:
I see, so just these loads can be relaxed ordering. I agree then.